packages/cli/src/UserManagement/decorators.ts

```diff
@@ -0,0 +1,89 @@
+/* eslint-disable @typescript-eslint/no-unsafe-assignment */
+/* eslint-disable @typescript-eslint/naming-convention */
+/* eslint-disable import/no-cycle */
+import { Application, Request, RequestHandler as Middleware, Response, Router } from 'express';
+import { posix } from 'path';
+import { ResponseHelper } from '..';
+import config from '../../config';
+
+interface RequestHandler<ReqBody, ResBody> extends Function {
+	async(req: Request<unknown, ResBody, ReqBody>, res: Response<ResBody>): Promise<ReqBody>;
+}
+
+export interface IRouter {
+	method: Method;
+	path: string;
+	handlerName: string | symbol;
+}
+
+const Keys = {
+	ROUTES: Symbol('routes'),
+	BASE_PATH: Symbol('base_path'),
+} as const;
+
+export enum Method {
+	GET = 'get',
+	POST = 'post',
+	PATCH = 'patch',
+	DELETE = 'delete',
+}
+
+export const RestController =
+	(basePath = '/'): ClassDecorator =>
+	(target: object) => {
+		Reflect.defineMetadata(Keys.BASE_PATH, basePath, target);
+	};
+
+const RouteFactory =
+	(...methods: Method[]) =>
+	(path = '/'): MethodDecorator =>
+	(target, handlerName) => {
+		const ControllerClass = target.constructor;
```

Comment:
If the metadata holder can grab a reference to its containing module instead, we could think about doing away with the class? Making all decorated methods into standalone decorated functions. The shared path segment would become part each decorator argument, which would be better for searchability as opposed to being split up.

Reply:
I'm not sure how that would work. could you please share a code snippet to clarify?

Reply:
```ts
const ControllerClass = target.constructor;
```

I notice that `target` can grant access to its containing class when `target` is a method, so assuming `target` can grant access to the containing _module_ in a similar way, then it strikes me that the class itself would be unneeded as a container and request handlers could be set at the top level of the module, with the shared segments that are now args to `@RestController` being part of the function args - to make the endpoints more searchable and the setup simpler.

```ts
// standalone function, not class method
@POST('/users')
async function createUser() {
	// ...
}
```
I learned about `reflect-metadata` today so I could be way off base. I trust your opinion as to whether this is possible.

Reply:
from what I understand TS decorators can only be applied on and inside classes. That's the reason why I had to create these classes, instead of using a simple object with methods, or using functions directly.

Reply:
Thank you, then definitely not possible.